Parse externally_connectable properties from the manifest

kiaraarose opened the pull request:
#### 47b3bd6baca3b003554ea69f8bb809307fd999b4
<pre>
Parse externally_connectable properties from the manifest
<a href="https://bugs.webkit.org/show_bug.cgi?id=246491">https://bugs.webkit.org/show_bug.cgi?id=246491</a>
<a href="https://rdar.apple.com/114823306">rdar://114823306</a>

Reviewed by Timothy Hatcher.

Parse externally_connectable properties from the manifest.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::populateExternallyConnectableIfNeeded):
(WebKit::WebExtension::externallyConnectableMatchPatterns):
(WebKit::WebExtension::allRequestedMatchPatterns):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: <a href="https://commits.webkit.org/274665@main">https://commits.webkit.org/274665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c0460f2fe3cae3b851f8ec48a1df78b14c1d056

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18685 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42062 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42241 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42013 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21586 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16014 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40280 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15762 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13663 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13659 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35306 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43518 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36052 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35639 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39423 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14518 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11956 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16124 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16173 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5215 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15781 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -843,6 +848,65 @@ static _WKWebExtensionError toAPI(WebExtension::Error error)
+            if (!hostWithTopLevelDomainRemoved.length || [hostWithTopLevelDomainRemoved isEqualToString:@"*"]) {
+                shouldReportError = true;
+                continue;
+            }
```
I think you can do this:

```c++
#import <WebCore/PublicSuffix.h>
```

```c++
// URL patterns must contain at least a second-level domain. Top level domains and wildcards are not standalone patterns.
String host = matchPattern->host();
if (host.startsWith("*."_s))
    host = host.substring(2);

if (isPublicSuffix(host)) {
    shouldReportError = true;
    continue;
}
```

xeenon replied:
You could make this be a helper method on `WebExtensionMatchPattern` too!

kiaraarose replied:
thank you!

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -276,6 +276,10 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
     const MatchPatternSet& requestedPermissionMatchPatterns();
     const MatchPatternSet& optionalPermissionMatchPatterns();
 
+    // Permission patterns requested by the extension in their manifest.
```
Not `Permission patterns…`. `Match patterns…`?

kiaraarose committed:
Parse externally_connectable properties from the manifest
https://bugs.webkit.org/show_bug.cgi?id=246491
rdar://114823306

Reviewed by Timothy Hatcher.

Parse externally_connectable properties from the manifest.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::populateExternallyConnectableIfNeeded):
(WebKit::WebExtension::externallyConnectableMatchPatterns):
(WebKit::WebExtension::allRequestedMatchPatterns):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: https://commits.webkit.org/274665@main
